pagopa/io-app
# [#172889149] Automatic changelog and versioning generation

**Short description:**
This pr allows to versioning the code, creating a "release" and automatically generate a changelog for that release.

The new app version (based on [Semantic Versioning](https://semver.org/)) is automatically calculated based on the utility [Standard Version](https://github.com/conventional-changelog/standard-version). Using in the commit title the tags `feat: `, `fix: ` or `chore: ` it will be possible use this information to calculate the new version number and the changelog.

### How to format a squash&merge commit
In order to automatically calculate the next application version and generate the changelog, each squash&merge commit must respect the following syntax:

`fix|feat|chore: [#pivotal_id] text of the commit (#pull_request_id)` 

for example, for this pr, it would be necessary add the tag `chore: ` to the default squash commit in order to register the title in the changelog.

### How to create a release
Using the command `yarn release` or `yarn release:rc` (to generate a relaese candidate, with the `-rc.x`) will triggers the following steps:

1) Check if the Git working tree is clean
2) Calculate the **next version number** and use this value to update:
a) `version` in `package.json`
b) `CFBundleShortVersionString` in `ios/ItaliaApp/Info.plist`
c) `CFBundleShortVersionString` in `ios/ItaliaAppTests/Info.plist`
d) `versionName` in `android/app/build.gradle`
3) Increase the *build incremental number* for:
a) `CURRENT_PROJECT_VERSION` in `ios/ItaliaApp.xcodeproj/project.pbxproj`
b) `CFBundleVersion` in `ios/ItaliaApp/Info.plist`
c) `CFBundleVersion` in `ios/ItaliaAppTests/Info.plist`
d) `versionCode` in `android/app/build.gradle`
4) Generate changelog
5) Commit modified files
6) Apply a tag with the **next version number**

It's possible to call the commands `yarn release` and `yarn release:rc` with the argument `--dry-run` in order to have a preview of the **next version number**  and the **Changelog** without modifying or commit any file (only to console).

**List of changes proposed in this pull request:**
- Added the utility `standard-version` in dev dependencies with all the customization for this project.
- Added the commands `yarn release` and `yarn release:rc`.
- Changed `android/app/build.gradle` in order to be updated from the release phase instead that using the `CIRCLE_BUILD_NUM`
- Changed `ios/fastlane/Fastfile` in order to avoid to incraese the build version each time and commit files that aren't changed anymore.

**How to test:**
Create some local commits with the title `feat: A new feature`, `fix: A fix` and so on then launch `yarn release` to verify that all the file are updated correcly, the changelog is generated, the next version number is calculated correcly and the tag on the git repository is created.
